cloud_storage: rapid_json throws bad_alloc

LenaAn opened the pull request:
## Cover letter

This PR add a wrapper around rapid_json memory allocator that throws bad_alloc if memory was requested, but wasn't allocated. cloud_storage manifests now use this allocator when parsing json document.

## Release notes
* none

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+#include <rapidjson/document.h>
+#include <rapidjson/encodings.h>
+
+namespace rjson_allocator {
```
I think this file should be named `src/v/json/allocator.h` and the namespace `json`.

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+#include <rapidjson/encodings.h>
+
+namespace rjson_allocator {
+class MyAllocator : public rapidjson::CrtAllocator {
```
Can this should have a more informative name and be `snake_case`.

I think it would be better to implement concept directly and reuse `CrtAllocator` by composition and delegation rather than inheritance.

BenPope replied:
Althought you may want to use [\[\[no_unique_address\]\] _impl;](https://en.cppreference.com/w/cpp/language/attributes/no_unique_address)

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+class MyAllocator : public rapidjson::CrtAllocator {
+public:
+    void* Malloc(size_t size) {
+        if (!size) {
+            return NULL;
+        }
```
This is already checked by `CrtAllocator`, it's probably best to defer responsibility.

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+        }
+
+        void* res = rapidjson::CrtAllocator::Malloc(size);
+        if (NULL == res) {
```
Prefer `nullptr`, or simply `!res`.

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+
+        return res;
+    }
+};
```
I think `CrtAllocator::Realloc` returns null on failure, should that be checked, too?

LenaAn replied:
added a wrapper for `CrtAllocator::Realloc`

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+
+        void* res = rapidjson::CrtAllocator::Malloc(size);
+        if (NULL == res) {
+            throw std::bad_alloc{};
```
Can we add a `fmt_with_ctx` here to get a stack trace.

BenPope commented on src/v/utils/rjson_allocator.h:
```diff
@@ -0,0 +1,26 @@
+    }
+};
+
+using Document = rapidjson::GenericDocument<
```
This probably doesn't belong in a file named "allocator".

Are there any other types where the allocator should be replaced?

LenaAn replied:
moved it to the `document.h`

for other types I think we can do it in the separate PR

jcsp commented:
Not much to add to Ben's comments -- I think the main things are to refactor into `src/v/json` and cover the realloc case.  I'm neutral on the inheritance vs. encapsulation thing.

BenPope commented on src/v/json/allocator.h:
```diff
@@ -0,0 +1,27 @@
+#include <rapidjson/allocators.h>
+namespace json {
+class throwing_allocator : public rapidjson::CrtAllocator {
```
There's now a `CrtAllocator` base, and member (they're stateless, but it's confusing).

I think remove the base and implement `Free`.

LenaAn replied:
oh good catch!

BenPope commented on src/v/json/allocator.h:
```diff
@@ -0,0 +1,27 @@
+    }
+
+private:
+    [[no_unique_address]] rapidjson::CrtAllocator rp_allocator;
```
Private members usually have a `_` prefix in our codebase.

BenPope reviewed (commented):
> adding fmt_with_ctx makes sense to me, but I can't find a way to provide a message to bad_alloc :(

I was thinking something like:
```
class bad_alloc_with_msg : public std::bad_alloc {
public:
    bad_alloc_with_msg(std::string msg) : _msg(std::move(msg)) {}
    const char* what() const noexcept override { return _msg.c_str(); }
private:
    std::string _msg;
};

jcsp reviewed (dismissed):
:+1:

jcsp commented:
release-clang-amd64 failure was unrelated (RaftAvailabilityTest.test_one_node_down)

senior7515 commented on src/v/json/allocator.h:
```diff
@@ -0,0 +1,48 @@
+    const char* what() const noexcept override { return _msg.c_str(); }
+
+private:
+    std::string _msg;
```
we don't use `std::string` we use `ss::string`

senior7515 replied:
are these strings really dynamic? or are they all defined up front in code. if they are just regular static strings, you can avoid the allocation all together. by returning a static string.

BenPope replied:
`fmt_with_ctx` is usually used with `fmt::format`, which returns a dynamic `std::string`. I guess it should be possible to use `ssx::sformat`, and it looks like there are some usages in serde and scram.

I guess allocating any string during `bad_alloc` is quite possibly not going to work, which is the likely reason `std::bad_alloc` doesn't accept one.

dotnwat replied:
perhaps for bad_alloc_with_msg we should store a `const char*` and then use __FILE__ and __LINE__ etc... to build the format string a string literal to avoid the allocations.

dotnwat replied:
or maybe that would just give us the file/line of the header. just returning bad_alloc is probably good enough.

senior7515 commented on src/v/json/allocator.h:
```diff
@@ -0,0 +1,48 @@
+    void* Malloc(size_t size) {
+        void* res = _rp_allocator.Malloc(size);
+        if (!res && (0 != size)) {
+            throw bad_alloc_with_msg(
```
this is fine, but you can remove all allocations if you log this and return a static string. just an idea

senior7515 commented on src/v/json/document.h:
```diff
@@ -0,0 +1,11 @@
+#include <rapidjson/encodings.h>
+
+namespace json {
+using Document = rapidjson::GenericDocument<
```
Oh this is neat!

dotnwat reviewed (commented):
everything looks good. avoiding the string formatting when throwing std::bad_alloc makes sense to change.

LenaAn commented:
in [force push](https://github.com/vectorizedio/redpanda/compare/9515b5d60131a5c9a6846f6293489bb9f83d1391..f16302de2950a11ca9e0347d4da04a56a98bfd45): 
 - ended up just logging an error and throwing `std::bad_alloc`

BenPope reviewed (commented):
Looks pretty good to me.

It might be worth defining a `json::Writer`, since `rapidjson::Writer` allocates.

The new files are missing the copyright notice.

LenaAn commented:
> It might be worth defining a `json::Writer`, since `rapidjson::Writer` allocates.

Will do in a separate PR, created https://github.com/vectorizedio/redpanda/issues/3561 to track that 

> The new files are missing the copyright notice.

Added

BenPope reviewed (changes requested):
Can the new files be licenced under the BSL, so that they can be used in BSL parts of the codebase.

Without #3561 I think https://github.com/vectorizedio/redpanda/issues/3429 cloud_storage will still segfault on low memory, just with a difference call stack.

LenaAn commented:
> Can the new files be licenced under the BSL, so that they can be used in BSL parts of the codebase.

Fixed. And what are the terms for the license that I used before?

> _Note_ (and this can wait for a different PR) Without #3561 I think #3429 cloud_storage will still segfault on low memory, just with a difference call stack.

Deleted the FIXES part in the PR description

BenPope reviewed (approved):
LGTM
